Fix #9924, prefill prefix and suffix in checkout shipping address

AntonEvers opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->
Prefill prefix and suffix in the checkout if they are present on a logged in customer.

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#9924 : Prefix and suffix are not prefilled in the quote shipping address

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. start an order, using an existing account with a prefix and a suffix
2. visit the shipping address step in the checkout
3. prefix and suffix should be prefilled in the shipping address form for customers without an address
4. prefix and suffix should be prefilled in the NEW shipping address form for customers with addresses

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Fix #9924, prefill prefix and suffix in checkout shipping address

Commit:
Simplify the getDefaultValue method in an attempt to pass PHP Mess Detection checks

Commit:
Bring cyclomatic complexity down from 10 to 6

ishakhsuvarov commented on app/code/Magento/Checkout/Block/Checkout/AttributeMerger.php:
```diff
@@ -388,4 +387,36 @@ protected function orderCountryOptions(array $countryOptions)
+    /**
+     * @return null|string
+     */
+    protected function getDefaultPrefix()
```
Magento does not encourage usage of `protected` methods.

AntonEvers replied:
Good to know. Because of plugins right?

ishakhsuvarov replied:
Because we do not encourage inheritance-based API. So composition in favor of inheritance. It's easier to maintain and reuse.

AntonEvers replied:
Ah yes of course. That makes sense.

ishakhsuvarov commented on app/code/Magento/Checkout/Block/Checkout/AttributeMerger.php:
```diff
@@ -316,20 +316,19 @@ protected function getMultilineFieldConfig($attributeCode, array $attributeConfi
     protected function getDefaultValue($attributeCode)
     {
         switch ($attributeCode) {
```
Wouldn't it be better if you return `null` straight away, at the beginning of the method in case where there is no customer? I guess it should reduce complexity that way, since you would not have to check for customer every time.

AntonEvers replied:
Hahaha, brilliant... I guess programming for 12 hours straight is not paying off ;)

AntonEvers replied:
Correction, `$this->directoryHelper->getDefaultCountry()` doesn't need a customer. Let me draft up a commit.

Commit:
Bring cyclomatic complexity down from 6 to 4

ajpevers commented:
@ishakhsuvarov how about this?

Commit:
ISSUE-9924 Merge remote-tracking branch 'upstream/develop' into issue-9924

* upstream/develop: (60 commits)
  Fix typo in comment
  Move prefix and suffix default values to a new PR
  MAGETWO-68877: Issue #7988 Typo changed also added comments for each index, getters and setters. #9484
  Revert "MAGETWO-69728: Fixes layered navigation options being cached using the wrong store id. #9873"
  MAGETWO-67500: setup:di:compile returns exit code 0 if errors are found #7780
  Fix prefix, middle name and suffix were not prefilled in the checkout
  add middle name to checkout address html templates #8878
  Using Command output as message which actually provides more information for debugging than just the path
  Handling CLI error as a failure when validating composer.json file
  MAGETWO-69805: Return array of blocks as items instead of array of arrays #9157
  MAGETWO-69666: Return array of pages as items instead of array of arrays #9823
  MAGETWO-69723: Email to a Friend feature #9824
  MAGETWO-69539: PHP "soap" extension is not declared in composer.json but can be used by Magento modules
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  MAGETWO-63054: [Catalog] MSRP field is not displayed for bundle products with fixed price
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  MAGETWO-67087: Saving category deletes url-rewrites for products in anchor categories
  ...
